### PR TITLE
WIP: Calculate argument buffer allocation size taking variable counts into account.

### PR DESCRIPTION
**Pull request by js6i**
When allocating descriptors from the pool, we should only request as much as needed for the variable number of descriptors, if applicable.

It looks slightly dicey to use the fixed size argument encoder when encoding variable sized bindings. Since here we're creating the encoder to calculate the size, maybe we could use it for encoding too..

### CODE REVIEW

File: MoltenVK/MoltenVK/GPUObjects/MVKDescriptorSet.mm
Comment (billhollings) [inline MoltenVK/MoltenVK/GPUObjects/MVKDescriptorSet.mm:279]: Since only the last descriptor binding in a descriptor set layout can be variable, instead of tracking `_hasVariableDescriptorCount`, querying the last binding (if at least one exists) by calling `hasVariableDescriptorCount()` is simple. 

With that idea, this could also test that `variableDescriptorCount` is less than `getDescriptorCount()` of that same binding, to determine if calculating with a temporary argument encoder is necessary.

**Review by billhollings — CHANGES_REQUESTED** [review]
Thanks for submitting this.

I've requested a couple of changes.

> It looks slightly dicey to use the fixed size argument encoder when encoding variable sized bindings. Since here we're creating the encoder to calculate the size, maybe we could use it for encoding too..

Good point. And good catch.

Having each descriptor set carry its own encoder sounds like a lot. Perhaps we could have the descriptor set defer to the layout's encoder, unless the variable descriptor count is less than the descriptor count of the variable-length binding.

However, we are going to move away from argument encoders very soon. So, I don't think we should bother making things that complex unless we actually determine it is causing a problem.

File: MoltenVK/MoltenVK/GPUObjects/MVKDescriptor.mm
Comment (js6i) [reply MoltenVK/MoltenVK/GPUObjects/MVKDescriptor.mm]: Yes, I think that works.

File: MoltenVK/MoltenVK/GPUObjects/MVKDescriptorSet.mm
Comment (js6i) [reply MoltenVK/MoltenVK/GPUObjects/MVKDescriptorSet.mm]: Good points, done.

**Comment by js6i** [comment]
> Having each descriptor set carry its own encoder sounds like a lot. Perhaps we could have the descriptor set defer to the layout's encoder, unless the variable descriptor count is less than the descriptor count of the variable-length binding.

I think we could do that. Are they expensive though? I would have thought it's fine to create them on demand, for single use.

> However, we are going to move away from argument encoders very soon. So, I don't think we should bother making things that complex unless we actually determine it is causing a problem.

It's causing a validation error. That said, there's also another one, when the shader has the maximum size array declared, but the buffer that gets bound is smaller. Let's hold off merging for now, I'll see what can be done about the other one.

**Comment by billhollings** [comment]
This PR is almost certainly subsumed or made obsolete by the recent PRs #2260 & #2268.

Closing this as a result. Please reopen if needed.

**Comment by billhollings** [comment]
This PR is almost certainly subsumed or made obsolete by the recent PRs #2260 & #2268.

Closing this as a result. Please reopen if needed.

**Comment by billhollings** [comment]
Replaced by PR #2273.

**Pull request closed without merging**
